app/models/article.rb

```diff
@@ -54,7 +55,8 @@ class Article < ApplicationRecord
            inverse_of: :commentable,
            class_name: "Comment"
 
-  validates :body_markdown, length: { minimum: 0, allow_nil: false }, uniqueness: { scope: %i[user_id title] }
+  validates :body_markdown, length: { in: 0..MAX_BODY_MARKDOWN_SIZE, allow_nil: false }
+  validates :body_markdown, uniqueness: { scope: %i[user_id title] }
```

Comment:
A uniqueness validation here may not be a good idea. When Rails invokes a uniqueness validation, it'll run a query like this:

```sql
SELECT 1 AS one
FROM "articles"
WHERE "articles"."body_markdown" = $1
LIMIT $2
```

There is no index on `body_markdown`, so it has to check the column data. Because `body_markdown` can be pretty large, its content will almost always be held in [TOAST tables](https://wiki.postgresql.org/wiki/TOAST), adding another place to look on disk during the query, making this slower than a simple value lookup.

To be clear, I'm also not saying we should add an index to this. Indexing `body_markdown` would bloat the indexes — on DEV, this column alone is over 2GB and all other indexes combined on this table is currently 3GB.

If we want to ensure that double-clicking the submit button doesn't create duplicate articles, I'd recommend creating a `body_markdown_checksum` column, storing a hash of the `body_markdown` in it, and creating a unique index on that column.

Reply:
@jgaskins thanks for the feedback, though I think everything you said has been already covered. Let me know if we should change something

A bit of context:

the uniqueness validator already exists in production, I only moved it to a different line for readability:

https://github.com/forem/forem/blob/c842d1a0c6718f319f167e9ec13d4d7d5d6b6d39/app/models/article.rb#L57

We do have an index on `body_markdown` which uses a digest of the markdown:

https://github.com/forem/forem/blob/c842d1a0c6718f319f167e9ec13d4d7d5d6b6d39/db/schema.rb#L155

Reply:
Apologies, I misread the diff. I didn't notice this was an existing validation. This is what I saw for the changed line 😆 

<img width="910" alt="Screenshot of the diff of the first line" src="https://user-images.githubusercontent.com/108205/115052902-7c2fc100-9eac-11eb-83a6-2bdbd865ddf4.png">

The line wrapping had me all confused. 😞 

Since it's already live, I looked at some traces where this happens and it's surprisingly fast (1-2ms), apparently because it's not globally unique like I thought when I read it. Still, I don't _love_ the idea of checking the database for uniqueness on something that can be 800KB since it negates any benefit we were gaining from just checking the digest for uniqueness, but that's not a problem for this PR to solve.

Sorry, my brain's apparently still fried from RailsConf. 😄